edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs

```diff
@@ -91,33 +91,77 @@ public async Task<IList<string>> GetAllIds()
             }
         }
 
-        public async Task<Option<string>> GetAuthChain(string id)
+        public async Task<Option<string>> GetAuthChain(string targetId)
         {
-            Preconditions.CheckNonWhiteSpace(id, nameof(id));
+            Preconditions.CheckNonWhiteSpace(targetId, nameof(targetId));
 
             // Auth-chain for the acting device (when there aren't any nodes in the tree yet)
-            if (id == this.actorDeviceId)
+            if (targetId == this.actorDeviceId)
             {
                 return Option.Some(this.actorDeviceId);
             }
-            else if (id == this.actorDeviceId + "/$edgeHub")
+            else if (targetId == this.actorDeviceId + "/" + Constants.EdgeHubModuleId)
             {
-                // TODO: richma - make this better
-                return Option.Some(this.actorDeviceId + "/$edgeHub" + ";" + this.actorDeviceId);
+                return Option.Some(this.actorDeviceId + "/" + Constants.EdgeHubModuleId + ";" + this.actorDeviceId);
             }
 
             using (await this.nodesLock.LockAsync())
             {
                 // Auth-chain for a child somewhere in the tree
-                if (this.nodes.TryGetValue(id, out ServiceIdentityTreeNode treeNode))
+                if (this.nodes.TryGetValue(targetId, out ServiceIdentityTreeNode treeNode))
                 {
+                    // Check every Edge device in the authchain for disabled devices
+                    string[] authChainIds = treeNode.AuthChain.Map(chain => chain.Split(new char[] { ';' }, StringSplitOptions.RemoveEmptyEntries)).OrDefault();
+
+                    foreach (string chainId in authChainIds)
```

Comment:
I suppose that this is what we discussed last time, but why not do that in the function validate auth chain? 
You have a loop very similar there.

Reply:
because this validation requires that we have access to the cache, so that we can check the individual identities.  The authchain validation function is really more of a format check, and doesn't contain any logic specific to the metadata in each identity of the link.

---
In reply to: [457535659](https://github.com/Azure/iotedge/pull/3248#discussion_r457535659) [](ancestors = 457535659)